crates/tedge_cli/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+impl Command for TEdgeCommand {
+    type Error = anyhow::Error; // .. either the top-level error type that accumulates all the sub-command error types, or just anyhow::Error
+
+    fn description(&self) -> String {
+        match self {
+            Self::CertCommand(cmd) => cmd.description(),
+            Self::GetConfigCommand(cmd) => cmd.description(),
+            Self::ConnectCommand(cmd) => cmd.description(),
+        }
+    }
+
+    fn execute(self) -> Result<(), Self::Error> {
+        match self {
+            Self::CertCommand(cmd) => cmd.execute(),
+            Self::GetConfigCommand(cmd) => cmd.execute(),
+            Self::ConnectCommand(cmd) => cmd.execute(),
+        }
+    }
+}
+ 
+impl TEdgeCli {
+    pub fn into_command(self, context: ...) -> Result<TEdgeCommand, CliError> {
+        match self.tedge_command {
+            TEdgeCommandCli::Cert(opt) => {
+                unimplemented!()
+                // ...
+            }
+            TEdgeCommandCli::Config(config_cmd) => {
+                match config_cmd {
+                    ConfigCommandCli::Get { key } => Ok(
+                        TEdgeCommand::GetConfigCommand(
+                        tedge_command_config::GetConfigCommand {
+                            config_key: key,
+                            config
+                        }))
+
+                    // ...
+                }
+            }
+            TEdgeOpt::Connect(opt) => {
+                unimplemented!()
+            }
+            TEdgeOpt::Disconnect(opt) => {
+                unimplemented!()
+            }
+            TEdgeOpt::Mqtt(opt) => {
+                unimplemented!()
+            }
+        }
+
+    }
+}
+
+// These kind of tests ensure that our CLI parser works correctly.
+#[test]
+fn test_cli_parsing() {
+    let cli: TEdgeCli = TEdgeCli::parse_from("tedge config set device.id 123")?; 
+    assert_matches!(cli, TEdgeCli {
+        tedge_command: TEdgeCommandCli::Config(
+                           ConfigCommandCli::Set {
+                               key: tedge_command_config::ConfigKey::from_str("device.id")?,
+                               value: "123",
+                           }
+        )
+    });
+}
+
+// These kind of tests ensure that given a TEdgeCli, we convert it into the correct commands.
+#[test]
+fn test_cli_into_command() {
+    let cli = TEdgeCli {
+        tedge_command: TEdgeCommandCli::Config(
+                           ConfigCommandCli::Set {
+                               key: tedge_command_config::ConfigKey::from_str("device.id")?,
+                               value: "123".into(),
+                           }
+        )
+    };
+
+    let command = cli.into_command();
+
+    assert_matches!(command, TEdgeCommand::SetConfigCommand {
+      key: tedge_command_config::ConfigKey::from_str("device.id")?,
+      value: "123".into(),
+      // ...
+    });
+    assert_matches!(command.description(), "....");
+}
+
+#[test]
+fn test_command_execution() {
```

Comment:
We don't run commands here! Just to mention, THESE KIND OF TESTS DO NOT BELONG HERE. They belong to either e2e tests or to the individual command implementations.